examples/lock-app/genio/src/LockManager.cpp

```diff
@@ -0,0 +1,707 @@
+
+    auto & userInStorage = mLockUsers[userIndex];
+
+    if (userName.size() > DOOR_LOCK_MAX_USER_NAME_SIZE)
+    {
+        ChipLogError(Zcl, "Cannot set user - user name is too long [endpoint=%d,index=%d]", endpointId, userIndex);
+        return false;
+    }
+
+    if (totalCredentials > LockParams.numberOfCredentialsPerUser)
+    {
+        ChipLogError(Zcl, "Cannot set user - total number of credentials is too big [endpoint=%d,index=%d,totalCredentials=%u]",
+                     endpointId, userIndex, totalCredentials);
+        return false;
+    }
+
+    strncpy(mUserNames[userIndex], userName.data(), userName.size());
+    userInStorage.userName       = chip::CharSpan(mUserNames[userIndex], userName.size());
+    userInStorage.userUniqueId   = uniqueId;
+    userInStorage.userStatus     = userStatus;
+    userInStorage.userType       = usertype;
+    userInStorage.credentialRule = credentialRule;
+    userInStorage.lastModifiedBy = modifier;
+    userInStorage.createdBy      = creator;
+
+    for (size_t i = 0; i < totalCredentials; ++i)
+    {
+        mCredentials[userIndex][i] = credentials[i];
+    }
+
+    userInStorage.credentials = chip::Span<const DlCredential>(mCredentials[userIndex], totalCredentials);
+
+    // Save user information in NVM flash
+    MT793XConfig::WriteConfigValueBin(MT793XConfig::kConfigKey_LockUser, reinterpret_cast<const uint8_t *>(&mLockUsers),
+                                      sizeof(EmberAfPluginDoorLockUserInfo) * LockParams.numberOfUsers);
+
+    MT793XConfig::WriteConfigValueBin(MT793XConfig::kConfigKey_UserCredentials, reinterpret_cast<const uint8_t *>(mCredentials),
+                                      sizeof(DlCredential) * LockParams.numberOfUsers * LockParams.numberOfCredentialsPerUser);
+
+    MT793XConfig::WriteConfigValueBin(MT793XConfig::kConfigKey_LockUserName, reinterpret_cast<const uint8_t *>(mUserNames),
+                                      sizeof(mUserNames));
+
+    ChipLogProgress(Zcl, "Successfully set the user [mEndpointId=%d,index=%d]", endpointId, userIndex);
+
+    return true;
+}
+
+bool LockManager::GetCredential(chip::EndpointId endpointId, uint16_t credentialIndex, DlCredentialType credentialType,
+                                EmberAfPluginDoorLockCredentialInfo & credential)
+{
+
+    if (DlCredentialType::kProgrammingPIN == credentialType)
+    {
+        VerifyOrReturnValue(IsValidCredentialIndex(credentialIndex, credentialType),
+                            false); // programming pin index is only index allowed to contain 0
+    }
+    else
+    {
+        VerifyOrReturnValue(IsValidCredentialIndex(--credentialIndex, credentialType), false); // otherwise, indices are one-indexed
+    }
+
+    ChipLogProgress(Zcl, "Lock App: LockManager::GetCredential [credentialType=%u], credentialIndex=%d",
+                    to_underlying(credentialType), credentialIndex);
+
+    const auto & credentialInStorage = mLockCredentials[credentialIndex];
```

Comment:
This is not spec-compliant.  Credential indices in the spec are per-type, not a flat space.

Reply:
Hi, @bzbarsky-apple, line 405~413 should have taken care of per-type indices.

Reply:
Those lines don't take care of it.   What happens if a GetCredential happens for index 1, type PIN, and then a GetCredential happens for index 1, type RFID?  They should return different credentials, if both PIN and RFID are supported.  

Or does this app claim to only support one type of non-programming-user credential?  Because the feature map I am seeing certainly has the RID bit set and nothing I see in this PR changes it....

> we leveraged EFR32's implementation as the base

That implementation is wrong, for a door lock that supports both PIN and RID.

Reply:
Thanks for explaining to us, I've added feature map setting and enable only PIN since we don't have plan to support RFID any time soon.